// check-lint/lib/config/eslint/rax.js
/**
 * rax 规则集继承 react 规则，有一些覆盖
 */

module.exports = {
  extends: 'eslint-config-sendinfo/react',
  parser: 'babel-eslint',
  settings: {
    react: {
      pragma: 'createElement',
    },
  },
  rules: {
    // 不要在 JSX 属性中使用 .bind()
    'react/jsx-no-bind': [
      'error',
      {
        ignoreRefs: false,
        allowArrowFunctions: true,
        allowBind: false,
      },
    ],

    // 由于这个bug https://github.com/babel/babel-eslint/issues/799#issuecomment-568195009
    // 关闭 template-curly-spacing，修改 indent
    'template-curly-spacing': 'off',
    indent: [
      'error',
      2,
      {
        SwitchCase: 1,
        VariableDeclarator: 1,
        outerIIFEBody: 1,
        // MemberExpression: null,
        FunctionDeclaration: {
          parameters: 1,
          body: 1,
        },
        FunctionExpression: {
          parameters: 1,
          body: 1,
        },
        CallExpression: {
          arguments: 1,
        },
        ArrayExpression: 1,
        ObjectExpression: 1,
        ImportDeclaration: 1,
        flatTernaryExpressions: false,
        ignoredNodes: [
          'TemplateLiteral',
          'JSXElement',
          'JSXElement > *',
          'JSXAttribute',
          'JSXIdentifier',
          'JSXNamespacedName',
          'JSXMemberExpression',
          'JSXSpreadAttribute',
          'JSXExpressionContainer',
          'JSXOpeningElement',
          'JSXClosingElement',
          'JSXText',
          'JSXEmptyExpression',
          'JSXSpreadChild',
        ],
        ignoreComments: false,
      },
    ],
  },
};
