// check-lint/lib/config/eslint/vue.js
module.exports = {
  extends: 'eslint-config-sendinfo/vue',
  parserOptions: {
    parser: 'babel-eslint',
  },
  rules: {
    // 这条规则会不明原因误报，暂关闭
    'vue/require-component-is': 'off',

    // 由于这个bug https://github.com/babel/babel-eslint/issues/799#issuecomment-568195009
    // 关闭 template-curly-spacing，修改 indent
    'template-curly-spacing': 'off',
    indent: [
      'error',
      2,
      {
        SwitchCase: 1,
        VariableDeclarator: 1,
        outerIIFEBody: 1,
        // MemberExpression: null,
        FunctionDeclaration: {
          parameters: 1,
          body: 1,
        },
        FunctionExpression: {
          parameters: 1,
          body: 1,
        },
        CallExpression: {
          arguments: 1,
        },
        ArrayExpression: 1,
        ObjectExpression: 1,
        ImportDeclaration: 1,
        flatTernaryExpressions: false,
        ignoredNodes: [
          'TemplateLiteral',
          'JSXElement',
          'JSXElement > *',
          'JSXAttribute',
          'JSXIdentifier',
          'JSXNamespacedName',
          'JSXMemberExpression',
          'JSXSpreadAttribute',
          'JSXExpressionContainer',
          'JSXOpeningElement',
          'JSXClosingElement',
          'JSXText',
          'JSXEmptyExpression',
          'JSXSpreadChild',
        ],
        ignoreComments: false,
      },
    ],
  },
};
